SCF/OOP_Attempt/Epack.py

```diff
@@ -0,0 +1,284 @@
+        ndocc = self.ndocc
+        nvirt = self.nvirt
+
+        #Initial Diagonalization
+        C, Cocc, Cvirt, D = self.diagonalize(self.H,A)
+        
+        #Size Check
+        size = (nbf**4)*8/1.e9
+        if size > 14:
+            raise Exception("G tensor is too big!")
+
+        #Accessing Qpq through Psi4
+        aux = self.psi4.BasisSet.pyconstruct_auxiliary(self.wfn.molecule(), 'DF_BASIS_SCF', '', 'JKFIT', self.psi4.get_global_option('BASIS'))
+        zeroC = self.psi4.Matrix(nbf, nbf)
+        dfobj = self.psi4.DFTensor(self.wfn.basisset(), aux, zeroC, ndocc, nvirt)
+        Qpq = np.asarray(dfobj.Qso())
+
+        #Apply variational method which knows when to stop.
+        SCFE = 0
+        SCFE_old = 1
+        count = 0
+        Iter_type = "CORE"
+        SOSCF = True
+
+        #Initiate set of trial vectors, error vectors.
+        trials = []
+        errors = []
+
+        print ("Total time taken for setup: %6.6f" % (time.time() - start))
+        itstart = time.time()
+
+        while count < self.maxmacro:
+            count += 1
+            
+            #Construct Fock Matrix and Compute Energy    
+            J, K = self.JK.build_DFJK(Qpq, Cocc)
+            F = self.H + 2*J - K
+
+            Eelec = np.einsum('qr,qr->', F + self.H, D)
+            SCFE_old = SCFE
+            SCFE = Eelec + self.Enuc
+
+            grad = np.dot(F, D).dot(self.S) - np.dot(self.S, D).dot(F)
+            r = np.dot(A.T, grad).dot(A)
+            dRMS = np.mean(r**2)**0.5
+
+            #print metrics 
+            if count < 10:
+                scount = '0'
+                scount += str(count)
+            else:
+                scount = str(count)
+            print ('SCF Iteration %s: E = %6.6f  dE = % 2.5E  dRMS = %2.5E , %s' % (scount, SCFE, (SCFE_old - SCFE), dRMS, Iter_type))
+
+            #Iteration Logic
+            if (SCFE_old - SCFE) < self.dE_tol and dRMS < self.dRMS_tol:
+                break
+
+            elif SOSCF and dRMS < 1e-2: #Use SOSCF
+                C, Cocc, Cvirt, D, SOSCF = self.SOSCF(C, F, Qpq, self.ndocc, self.nvirt)
+                Iter_type = 'SOSCF'
+
+            else: #Use DIIS
+                F = self.DIIS(trials, errors, F, r)
+                C, Cocc, Cvirt, D = self.diagonalize(F, A)
+                Iter_type = 'DIIS'
+        
+        #End of iterations
+        if count == self.maxmacro:
+            raise Exception("Maximum Number of SCF iterations reached! (%d)" % self.maxmacro)
+
+        SCF_time = time.time() - start
+        print ("\nTotal time in iterations: %6.3f" % (time.time() - itstart))
+        
+        print ("Time in microiteratoins: %6.3f" % self.microtime)
+        print ("Number of iterations: %d" % count)
+        
+        
+        print ("Macro + Micro iterations: %d\n" % (self.totmicro + count))
+        print ("SCF Energy: %12.10f" % SCFE)
+        print ("Runtime:      %6.6f\n" % SCF_time)
+
+        #Calculate MP2 adjustement if specified
```

Comment:
`MP2` should probably be a separate class or at the very least a separate function.